test/test_auth.py

```diff
@@ -201,3 +201,117 @@ def test_scram_both(bouncer):
 
     # SCRAM password in userlist.txt
     bouncer.test(dbname="p62", user="scramuser1", password="foo")
+
+
+@pytest.mark.skipif("WINDOWS", reason="Windows does not have SIGHUP")
+def test_connection_with_missing_stats_user_and_enabled_auth_query_without_global_auth_dbname(
+    bouncer,
+):
+    """
+    Regression for https://github.com/pgbouncer/pgbouncer/issues/314
+
+    Check that pgbouncer is not crashing on connection to the virtual pgbouncer virtual db when user is missing in
+    userlists.txt and auth_query enabled.
+
+    It covers case when global auth_dbname does not set and target database used for auth_dbname in auth query
+    """
+
+    config = f"""
+        [databases]
+        * = host={bouncer.pg.host} port={bouncer.pg.port}
+        [pgbouncer]
+        auth_query = SELECT usename, passwd FROM pg_shadow where usename = $1
+        auth_user = pswcheck
+        stats_users = stats
+        listen_addr = {bouncer.pg.host}
+        verbose = 2
+        admin_users = pswcheck
+        auth_file = userlist.txt
+        listen_port = {bouncer.port}
+    """
+
+    # good password
+    bouncer.test(user="pgbouncer", password="fake")
+
+    with bouncer.run_with_config(config):
+        with pytest.raises(psycopg.OperationalError, match="bouncer config error"):
+            # bad password
+            # We expect that stats user does not exist ether in userlist.txt and in postgresql DB
+            bouncer.test(user="stats", password="stats", dbname="pgbouncer")
+
+
+@pytest.mark.skipif("WINDOWS", reason="Windows does not have SIGHUP")
+def test_connection_with_missing_stats_user_and_enabled_auth_query_with_global_auth_dbname(
```

Comment:
I think we should also add a test where auth_dbname is set correctly and `stats` user is not listed in `userlist.txt` file but exists in PostgreSQL. In this case, authentication should work when the target login database is `pgbouncer`

Reply:
That's a significant remark I will add this test

Reply:
Added tests that I hope cover all cases